### PR TITLE
Fixed #24210 Cleaned up relational fields __init__().

### PR DESCRIPTION
**Pull request by loic**

### CODE REVIEW

File: django/contrib/contenttypes/fields.py [hunk 1]
```diff
@@ -263,30 +270,31 @@ class GenericRelation(ForeignObject):
...
+
         # This construct is somewhat of an abuse of ForeignObject. This field
         # represents a relation from pk to object_id field. But, this relation
         # isn't direct, the join is generated reverse along foreign key. So,
         # the from_field is object_id field, to_field is pk because of the
         # reverse join.
-        super(GenericRelation, self).__init__(
-            to, to_fields=[],
-            from_fields=[self.object_id_field_name], **kwargs)
+        super(GenericRelation, self).__init__(to, [], [object_id_field], rel=rel, **kwargs)
```
Comment (collinanderson) [inline django/contrib/contenttypes/fields.py]: Why not use kwargs here? Compatibility?

**Comment by loic** [comment]
@timgraham when you have some time could you take a look at this PR, it should be in a reviewable state now.

File: django/contrib/contenttypes/fields.py [hunk 2]
```diff
@@ -253,6 +253,18 @@ def __set__(self, instance, value):
         setattr(instance, self.cache_attr, value)
 
 
+class GenericRel(ForeignObjectRel):
+    def __init__(self, field, to, related_name=None, related_query_name=None, limit_choices_to=None):
+
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: I'd personally omit the newlines you've added before super (don't recall this is a pattern we've used elsewhere)

File: django/contrib/contenttypes/fields.py [hunk 3]
```diff
@@ -263,30 +275,32 @@ class GenericRelation(ForeignObject):
...
         # This construct is somewhat of an abuse of ForeignObject. This field
         # represents a relation from pk to object_id field. But, this relation
         # isn't direct, the join is generated reverse along foreign key. So,
         # the from_field is object_id field, to_field is pk because of the
         # reverse join.
         super(GenericRelation, self).__init__(
-            to, to_fields=[],
-            from_fields=[self.object_id_field_name], **kwargs)
+            to, from_fields=[], to_fields=[object_id_field], **kwargs)
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: The arguments passed to `from_fields` and `to_fields` have switched? If it's correct, do we need a test for this?

File: django/db/models/fields/related.py [hunk 4]
```diff
@@ -2114,25 +2152,30 @@ def __init__(self, to, db_constraint=True, swappable=True, **kwargs):
...
             # here to break early if there's a problem.
             to = str(to)
-        kwargs['verbose_name'] = kwargs.get('verbose_name', None)
-        kwargs['rel'] = ManyToManyRel(
+
+        if symmetrical is None:
+            symmetrical = (to == RECURSIVE_RELATIONSHIP_CONSTANT)
+
+        if through is not None:
+            assert db_table is None, \
```
Comment (timgraham) [inline django/db/models/fields/related.py]: personally in favor of longer lines vs. backslashes

File: django/contrib/contenttypes/fields.py (hunk 3 above)
Comment (loic) [inline django/contrib/contenttypes/fields.py]: Haven't switched, they are defined in that order in `ForeignObject`, I thought I'd order the kwargs accordingly in the call to `super`.

Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: I see before:
`to_fields=[], from_fields=[self.object_id_field_name]`
after:
`to_fields=[object_id_field], from_fields=[]`
I could very well be missing something...

Comment (loic) [inline django/contrib/contenttypes/fields.py]: Hang on, you are right... I provided these as args, then I changed it back to kwargs according to Collin's feedback, I must have messed up in the process.

File: django/db/models/fields/related.py (hunk 4 above)
Comment (loic) [inline django/db/models/fields/related.py]: I've put it on a single line in the latest commit but I prefer when the assertion is standalone, then the message is less distracting. What about parentheses?

I removed backslashes from the other asserts and replaced them with parentheses, but these are multilines so it's less debatable.

File: django/contrib/contenttypes/fields.py (hunk 2 above)
Comment (loic) [inline django/contrib/contenttypes/fields.py]: Fixed. Some of those constructors are quite complex so I thought I'd carefully isolate each parts, but it's true it's quite foreign to this codebase and it looked funny in the few constructor that didn't really benefit from it.

File: django/db/models/fields/related.py (hunk 4 above)
Comment (timgraham) [inline django/db/models/fields/related.py]: parentheses are also fine with me.

**Pull request merged**